sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java

```diff
@@ -322,34 +317,40 @@ public Mono<DownloadBlobResult> downloadBlob(String digest) {
+    private Mono<BlobDownloadAsyncResult> downloadBlobInternal(String digest, Context context) {
         if (digest == null) {
             return monoError(LOGGER, new NullPointerException("'digest' can't be null."));
         }
 
-        return this.blobsImpl.getBlobWithResponseAsync(repositoryName, digest, context).flatMap(streamResponse -> {
-            String resDigest = UtilsImpl.getDigestFromHeader(streamResponse.getHeaders());
-
-            BinaryData binaryData = streamResponse.getValue();
-
-            // The service wants us to validate the digest here since a lot of customers forget to do it before consuming
-            // the contents returned by the service.
-            if (Objects.equals(resDigest, digest)) {
-                Response<DownloadBlobResult> response = new SimpleResponse<>(
-                    streamResponse.getRequest(),
-                    streamResponse.getStatusCode(),
-                    streamResponse.getHeaders(),
-                    new DownloadBlobResult(resDigest, binaryData));
-
-                return Mono.just(response);
-            } else {
-                return monoError(LOGGER, new ServiceResponseException("The digest in the response does not match the expected digest."));
-            }
-        }).onErrorMap(UtilsImpl::mapException);
+        Flux<ByteBuffer> content =
+            blobsImpl.getChunkWithResponseAsync(repositoryName, digest, new HttpRange(0, (long) CHUNK_SIZE).toString(), context)
+                .flatMapMany(firstResponse -> getAllChunks(firstResponse, digest, context))
+                .flatMapSequential(chunk -> chunk.getValue().toFluxByteBuffer(), 1);
+        return Mono.just(ConstructorAccessors.createBlobDownloadResult(digest, content));
+    }
+
+    private Flux<ContainerRegistryBlobsGetChunkResponse> getAllChunks(ContainerRegistryBlobsGetChunkResponse firstResponse, String digest, Context context) {
+        String responseHeaderDigest = firstResponse.getHeaders().getValue(DOCKER_DIGEST_HEADER_NAME);
+        if (!digest.equals(responseHeaderDigest)) {
+            return Flux.error(LOGGER.atError()
+                .addKeyValue("requestedDigest", digest)
+                .addKeyValue("responseDigest", responseHeaderDigest)
+                .log(new ServiceResponseException("The digest in the response header does not match the expected digest.")));
+        }
+
+        long blobSize = getBlobSize(firstResponse.getHeaders().get(HttpHeaderName.CONTENT_RANGE));
+        List<Mono<ContainerRegistryBlobsGetChunkResponse>> others = new ArrayList<>();
+        others.add(Mono.just(firstResponse));
+        for (long p = firstResponse.getValue().getLength(); p < blobSize; p += CHUNK_SIZE) {
+            HttpRange range = new HttpRange(p, (long) CHUNK_SIZE);
+            others.add(blobsImpl.getChunkWithResponseAsync(repositoryName, digest, range.toString(), context));
```

Comment:
Does Container Registry support HTTP preconditions? If so, we should pass a `If-Match: firstResponse.ETAG_HEADER` to ensure that we fail the download if the content changes between requests.

Reply:
the content is immutable  - the digest is sha256 of the content and it's used to get the content. We also validate it at the end.